python/pysrc/kaskada/windows.py

```diff
@@ -55,9 +62,73 @@ def yearly() -> Since:
         return Since(predicate=lambda domain: Timestream._call("yearly", domain))
 
 
+@dataclass(frozen=True)
+class Tumbling(Window):
+    """Window producing a single value at the time of a predicate.
+
+    "Tumbling" windows are a series of non-overlapping windows producing a single value
```

Comment:
Close. It seems like we're missing a detail -- "producing a single value at the time the predicate evaluates to true".

1. What happens if the predicate evaluates to true multiple times
2. How do we determine what values are included?

Maybe this is similar to Since windows? We need to do something to explain the window is between two successvie evaluations of the predicate to true.

Also note: This isn't *truly* tumbling windows. With tumbling windows the window is always the same size, and once it "ends" it just flips/shifts over and starts again. Specifically, with a predicate, these are likely different sizes. Even with something like "monthly", the "tumbling" equivalent would be fixed size (30 days or 31 days, but never changing) while our "monthly" is more of a "calendar monthly window".

Reply:
Yeah, that's true. Maybe we shouldn't use `tumbling` then, since it deviates from the "standard" definition.

Reply:
Yeah, but I'm not sure what we *should* call them then.

1. In Dataflow, they had "Calendar Windows" which makes sense for things like monthly. It's a bit odd for "hourly", but it can still (somewhat) make sense when you factor in leap seconds and such.
2. We could maybe call them tumbling if we had a way of explaining the difference. Perhaps it's something like "Dynamic Tumbling" or "Calendar Tumbling", where the property is that they are non-overlapping and when one window ends the next starts. That would let us have "Fixed tumbling" (exactly 3600 seconds for an hour) in the future.
3. These also feel kind of like a "session" between a predicate.